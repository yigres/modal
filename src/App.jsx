import React, { useState } from 'react';
import { useImmer } from 'use-immer';
import 'bootstrap/dist/css/bootstrap.css';

import Task from './Task';
import AddModal from './modals/AddModal';
import RemoveModal from './modals/RemoveModal';
import RenameModal from './modals/RenameModal';

const App = () => {
  const [showAdd, setShowAdd] = useState(false);
  const [renameId, setRenameId] = useState(null);
  const [removeId, setRemoveId] = useState(null);
  const [tasks, setTasks] = useImmer(['First Task!', 'Second Task!']);

  const taskToRename = tasks[renameId] || '';
  const taskToRemove = tasks[removeId] || '';

  const addTask = (task) => {
    setTasks((draft) => {
      draft.push(task);
    });
  };

  const renameTask = (task) => {
    setTasks((draft) => {
      draft[renameId] = task;
    });
  };

  const removeTask = () => {
    setTasks((draft) => {
      delete draft[removeId];
    });
  };

  const hideAddModal = () => setShowAdd(false);
  const hideRenameModal = () => setRenameId(null);
  const hideRemoveModal = () => setRemoveId(null);

  const showAddModal = () => setShowAdd(true);
  const createShowRenameModal = (id) => () => setRenameId(id);
  const createShowRemoveModal = (id) => () => setRemoveId(id);

  return (
    <>
      <div className="mb-3">
        <button type="button" onClick={showAddModal} data-testid="item-add" className="btn btn-secondary">add</button>
      </div>
      <AddModal show={showAdd} onHide={hideAddModal} addTask={addTask} />
      <RemoveModal onHide={hideRemoveModal} task={taskToRemove} removeTask={removeTask} />
      <RenameModal onHide={hideRenameModal} task={taskToRename} renameTask={renameTask} />
      {tasks.map((task, i) => (
        <Task
          key={task}
          title={task}
          remove={createShowRemoveModal(i)}
          rename={createShowRenameModal(i)}
        />
      ))}
    </>
  );
};

export default App;
