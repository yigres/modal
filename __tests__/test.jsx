import React from 'react';
import { waitForElementToBeRemoved, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import App from '../src/App';

describe('App', () => {
  it('application', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('application 2', async () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
    // asdf();

    const addButton = screen.getByTestId('item-add');
    userEvent.click(addButton);
    // ***********
    // console.log(addButton);
    const fr = asFragment();
    // console.log(Object.keys(fr));
    // console.log(screen.getByTestId('input-body'));
    // ***************
    expect(fr).toMatchSnapshot();

    const addingSubmitButton = screen.getByText('submit');
    const addingFormInput = screen.getByTestId('input-body');
    userEvent.type(addingFormInput, 'first task!');
    userEvent.click(addingSubmitButton);
    await waitForElementToBeRemoved(addingSubmitButton);
    expect(asFragment()).toMatchSnapshot();

    const renameLink = screen.getByTestId('item-rename');
    userEvent.click(renameLink);
    expect(asFragment()).toMatchSnapshot();

    const renamingFormInput = screen.getByTestId('input-body');
    const renamingSubmitButton = screen.getByText('submit');
    userEvent.type(renamingFormInput, 'changed name!');
    userEvent.click(renamingSubmitButton);
    await waitForElementToBeRemoved(renamingSubmitButton);
    expect(asFragment()).toMatchSnapshot();

    const removeLink = screen.getByTestId('item-remove');
    userEvent.click(removeLink);
    expect(asFragment()).toMatchSnapshot();

    const removingSubmitButton = screen.getByText('remove', { selector: 'input' });
    userEvent.click(removingSubmitButton);
    expect(asFragment()).toMatchSnapshot();
  });
});
