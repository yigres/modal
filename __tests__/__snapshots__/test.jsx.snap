// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App application 1`] = `
<DocumentFragment>
  <div
    class="mb-3"
  >
    <button
      class="btn btn-secondary"
      data-testid="item-add"
      type="button"
    >
      add
    </button>
  </div>
</DocumentFragment>
`;

exports[`App application 2 1`] = `
<DocumentFragment>
  <div
    class="mb-3"
  >
    <button
      class="btn btn-secondary"
      data-testid="item-add"
      type="button"
    >
      add
    </button>
  </div>
</DocumentFragment>
`;

exports[`App application 2 2`] = `
<DocumentFragment>
  <div
    class="mb-3"
  >
    <button
      class="btn btn-secondary"
      data-testid="item-add"
      type="button"
    >
      add
    </button>
  </div>
  <div
    class="modal-dialog"
  >
    <div
      class="modal-content"
    >
      <div
        class="modal-header"
      >
        <div
          class="modal-title h4"
        >
          Add
        </div>
        <button
          class="close"
          type="button"
        >
          <span
            aria-hidden="true"
          >
            ×
          </span>
          <span
            class="sr-only"
          >
            Close
          </span>
        </button>
      </div>
      <div
        class="modal-body"
      >
        <form>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              data-testid="input-body"
              name="body"
              required=""
              value=""
            />
          </div>
          <input
            class="btn btn-primary"
            type="submit"
            value="submit"
          />
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App application 2 3`] = `
<DocumentFragment>
  <div
    class="mb-3"
  >
    <button
      class="btn btn-secondary"
      data-testid="item-add"
      type="button"
    >
      add
    </button>
  </div>
  <div>
    <span
      class="mr-3"
    >
      first task!
    </span>
    <button
      class="border-0 btn-link mr-3 p-0"
      data-testid="item-rename"
      type="button"
    >
      rename
    </button>
    <button
      class="border-0 btn-link p-0"
      data-testid="item-remove"
      type="button"
    >
      remove
    </button>
  </div>
</DocumentFragment>
`;

exports[`App application 2 4`] = `
<DocumentFragment>
  <div
    class="mb-3"
  >
    <button
      class="btn btn-secondary"
      data-testid="item-add"
      type="button"
    >
      add
    </button>
  </div>
  <div>
    <span
      class="mr-3"
    >
      first task!
    </span>
    <button
      class="border-0 btn-link mr-3 p-0"
      data-testid="item-rename"
      type="button"
    >
      rename
    </button>
    <button
      class="border-0 btn-link p-0"
      data-testid="item-remove"
      type="button"
    >
      remove
    </button>
  </div>
  <div
    class="modal-dialog"
  >
    <div
      class="modal-content"
    >
      <div
        class="modal-header"
      >
        <div
          class="modal-title h4"
        >
          Rename
        </div>
        <button
          class="close"
          type="button"
        >
          <span
            aria-hidden="true"
          >
            ×
          </span>
          <span
            class="sr-only"
          >
            Close
          </span>
        </button>
      </div>
      <div
        class="modal-body"
      >
        <form>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              data-testid="input-body"
              name="body"
              required=""
              value="first task!"
            />
          </div>
          <input
            class="btn btn-primary"
            type="submit"
            value="submit"
          />
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App application 2 5`] = `
<DocumentFragment>
  <div
    class="mb-3"
  >
    <button
      class="btn btn-secondary"
      data-testid="item-add"
      type="button"
    >
      add
    </button>
  </div>
  <div>
    <span
      class="mr-3"
    >
      changed name!
    </span>
    <button
      class="border-0 btn-link mr-3 p-0"
      data-testid="item-rename"
      type="button"
    >
      rename
    </button>
    <button
      class="border-0 btn-link p-0"
      data-testid="item-remove"
      type="button"
    >
      remove
    </button>
  </div>
</DocumentFragment>
`;

exports[`App application 2 6`] = `
<DocumentFragment>
  <div
    class="mb-3"
  >
    <button
      class="btn btn-secondary"
      data-testid="item-add"
      type="button"
    >
      add
    </button>
  </div>
  <div>
    <span
      class="mr-3"
    >
      changed name!
    </span>
    <button
      class="border-0 btn-link mr-3 p-0"
      data-testid="item-rename"
      type="button"
    >
      rename
    </button>
    <button
      class="border-0 btn-link p-0"
      data-testid="item-remove"
      type="button"
    >
      remove
    </button>
  </div>
  <div
    class="modal-dialog"
  >
    <div
      class="modal-content"
    >
      <div
        class="modal-header"
      >
        <div
          class="modal-title h4"
        >
          Remove
        </div>
        <button
          class="close"
          type="button"
        >
          <span
            aria-hidden="true"
          >
            ×
          </span>
          <span
            class="sr-only"
          >
            Close
          </span>
        </button>
      </div>
      <div
        class="modal-body"
      >
        <form>
          <div
            class="form-group"
          >
            <input
              class="btn btn-danger"
              type="submit"
              value="remove"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App application 2 7`] = `
<DocumentFragment>
  <div
    class="mb-3"
  >
    <button
      class="btn btn-secondary"
      data-testid="item-add"
      type="button"
    >
      add
    </button>
  </div>
</DocumentFragment>
`;
